cockroachdb/cockroach
# logictest: refactor monolithic `TestLogic` targets into many subtests

The monolithic `TestLogic` test (and its corollaries, found in
`pkg/ccl/logictestccl` and `pkg/sql/opt/exec/execbuilder`) is very prone
to timing out under Bazel as the subtests cannot be parallelized by the
Bazel test runner. Furthermore, it can be difficult to debug failures in
individual subtests as the logs for each parallel subtest are all
intermingled. To solve this problem we generate finer-grained tests
for each of the subtests that would otherwise be run in the monolithic
`TestLogic` test. To that end we do all of the following:

1. Split out some common helper logic into the subpackage
   `pkg/sql/logictest/logictestbase`.
2. Refactor the logic test machinery. Up until this point, the
   `RunLogicTest` function (and related functions) have dynamically
   enumerated the subtests from `testdata` and spawned a subtest with
   `t.Run()` for each. Instead, `RunLogicTest` now takes a specific
   configuration index and a path to one test file.
3. Restructure the logictests as follows: split out the long-running
   sqlite logic tests into their own packages, `pkg/sql/sqlitelogictest`
   and `pkg/ccl/sqlitelogictestccl`; for each logictest configuration X,
   create a subpackage `tests/X` to house all tests for that
   configuration; for each logictest file Y, create a test
   `LogicTest_X_Y` that tests that one specific file. We do this not
   only for `pkg/sql/logictest` but also for `pkg/ccl/logictestccl` and
   `pkg/sql/opt/exec/execbuilder`.
4. Create a helper binary that generates all of these test files and
   integrate into `bazel-generate.sh`.
5. Update `Makefile` and `dev` to change how they execute logic tests.
6. Update some TeamCity scripts to change how they execute logic tests.
   Update `bazci` accordingly.
7. Delete now-irrelevant TeamCity scripts from `build`.

This has some breaking changes for developers:
1. Developers using `make testlogic` or `dev testlogic` should generally
   not be affected as these frontends are ported to the new test API's.
2. Developers used to running
     `dev test pkg/sql/logictest -f TestLogic/config-name/file-name/subtest`
   or the equivalent
     `make test ./pkg/sql/logictest TESTS=TestLogic/config-name/file-name/subtest`
   will now instead have to run:
     `dev test pkg/sql/logictest/tests/config-name -f TestLogic_file_name/subtest`
   or
     `make test ./pkg/sql/logictest/tests/config-name TESTS=TestLogic_file_name/subtest`
   In particular we point to the following important differences:
   * In the `file-name`, replace - and . with an underscore (_). If the
     filename has a slash, omit it entirely.
   * Use underscores instead of / in between the function name prefix,
     and the file name. All of these are baked into the actual Go test
     (function) name so we need to use _.
   * Don't include the config-name in the test filter; it's redundant.
   Note that if you just want to test all the tests in a directory with
   a particular config, the test filter is no longer necessary, so you
   can just do e.g. `dev test pkg/sql/logictest/tests/local`. The filter
   is necessary only if you want to test a particular file or subtest.
3. The same as above applies for logic tests in `pkg/ccl/logictestccl` and
   `pkg/sql/opt/exec/execbuilder`, as in
   `dev test pkg/sql/opt/exec/execbuilder/tests/local -f TestExecBuild_autocommit`.
4. Sqlite logic tests are split out into their own directory
   (`pkg/sql/sqlitelogictest` and `pkg/ccl/sqlitelogictestccl`) with the
   same test layout as above, like
   `dev test pkg/sql/sqlitelogictest/tests/local ...`
5. `dev testlogic` will now unconditionally regenerate these test files
   before running the logic test, and `dev test` will do the same if the
   test is a logic test target. This behavior can be skipped with
   `--no-gen`.

Closes https://github.com/cockroachdb/cockroach/issues/77204.
Closes https://github.com/cockroachdb/cockroach/issues/77456.

Release note: None
